File: apps/api/src/lib/seed.ts
/**
 * Idempotent seeding of the permission catalog and the default system roles
 * (SRS §11.8, FR-1.4). Returns a map of role key → role id.
 */
import { PERMISSIONS, DEFAULT_ROLE_LIST } from "@tuition/shared";
import type { Db } from "../types";
import { newId, nowIso } from "./id";

export async function seedPermissionsAndRoles(db: Db): Promise<Map<string, string>> {
  const now = nowIso();
  const stmts: D1PreparedStatement[] = [];

  // 1. Permission catalog (idempotent).
  for (const p of PERMISSIONS) {
    stmts.push(
      db
        .prepare(`INSERT OR IGNORE INTO permissions (key, resource, action, label) VALUES (?, ?, ?, ?)`)
        .bind(p.key, p.resource, p.action, p.label),
    );
  }

  // 2. Default roles — reuse existing ids where present.
  const existing = await db.prepare(`SELECT id, key FROM roles`).all<{ id: string; key: string }>();
  const roleIdByKey = new Map<string, string>();
  for (const r of existing.results ?? []) roleIdByKey.set(r.key, r.id);

  for (const role of DEFAULT_ROLE_LIST) {
    let id = roleIdByKey.get(role.key);
    if (!id) {
      id = newId("rol");
      roleIdByKey.set(role.key, id);
      stmts.push(
        db
          .prepare(
            `INSERT INTO roles (id, key, label, description, system, created_at, updated_at)
             VALUES (?, ?, ?, ?, 1, ?, ?)`,
          )
          .bind(id, role.key, role.label, role.description, now, now),
      );
    }
    for (const perm of role.permissions) {
      stmts.push(
        db
          .prepare(`INSERT OR IGNORE INTO role_permissions (role_id, permission_key) VALUES (?, ?)`)
          .bind(id, perm),
      );
    }
  }

  await db.batch(stmts);
  return roleIdByKey;
}

File: apps/api/internal/rbac/seed.go
package rbac

import (
	"context"
	"database/sql"
	"fmt"

	"attendly/api/internal/cryptox"
)

// Seed idempotently inserts the permission catalog and the default system
// roles (with their permission grants), reusing existing role ids where
// present. It returns a map of role key → role id.
func Seed(ctx context.Context, db *sql.DB) (map[string]string, error) {
	tx, err := db.BeginTx(ctx, nil)
	if err != nil {
		return nil, err
	}
	defer func() { _ = tx.Rollback() }()

	for _, p := range Permissions {
		if _, err := tx.ExecContext(ctx,
			`INSERT OR IGNORE INTO permissions (key, resource, action, label) VALUES (?, ?, ?, ?)`,
			p.Key, p.Resource, p.Action, p.Label,
		); err != nil {
			return nil, fmt.Errorf("seed permission %s: %w", p.Key, err)
		}
	}

	idByKey := make(map[string]string)
	rows, err := tx.QueryContext(ctx, `SELECT id, key FROM roles`)
	if err != nil {
		return nil, fmt.Errorf("list roles: %w", err)
	}
	for rows.Next() {
		var id, key string
		if err := rows.Scan(&id, &key); err != nil {
			_ = rows.Close()
			return nil, err
		}
		idByKey[key] = id
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	_ = rows.Close()

	now := cryptox.NowISO()
	for _, role := range DefaultRoles {
		id, ok := idByKey[role.Key]
		if !ok {
			id = cryptox.NewID("rol")
			idByKey[role.Key] = id
			if _, err := tx.ExecContext(ctx,
				`INSERT INTO roles (id, key, label, description, system, created_at, updated_at)
				 VALUES (?, ?, ?, ?, 1, ?, ?)`,
				id, role.Key, role.Label, role.Description, now, now,
			); err != nil {
				return nil, fmt.Errorf("seed role %s: %w", role.Key, err)
			}
		}
		for _, pk := range role.Permissions {
			if _, err := tx.ExecContext(ctx,
				`INSERT OR IGNORE INTO role_permissions (role_id, permission_key) VALUES (?, ?)`,
				id, pk,
			); err != nil {
				return nil, fmt.Errorf("grant %s to %s: %w", pk, role.Key, err)
			}
		}
	}

	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("commit seed: %w", err)
	}
	return idByKey, nil
}
